### DISCUSSION TITLE
Adjust log level based on http request parameter

### DISCUSSION THREAD

**Discussion by dominikwerder** [tokio-rs/tracing#3005] [Q&A] [answered]
In a rest service I would like to enable debug/trace log output for some certain modules based on the http query.

I have tried to `Instrument::instrument` the async request handler with a span like this:
```
if query.log_level() == TRACE {
    tracing::span!(tracing::Level::TRACE, "span_my_crate_trace")
} else {
    tracing::Span::none()
}
```

and then configure the EnvFilter with
```
info,my_crate[span_my_crate_trace]=trace
```

This basically works, but it now also enables trace events from other crates while the span is entered (notably tokio runtime).

How can I achieve this to see only the trace events from `my_crate` based on the query parameter, but not see other trace events from e.g. tokio?

**Comment by mladedav** [comment, marked as answer]
I don't think you can do this with a single `EnvFilter`. What I think should work though is using two filters, the first one the same as you're currently using and another one with `info,my_crate=trace`.

This should make it so both filters are queried and if both enable the the event, only then is it emitted.

The main problem here is that you cannot set both the filters with a single env variable (or you have to parse it yourself).

I am not entirely sure though because the dynamic span filters are a bit magic so please do challenge this if it doesn't work for you.

Ideally, we would probably want to support this in a single filter, but there is no syntax for this right as far as I know.

**Reply by dominikwerder** [reply]
This works really well!

Especially because I can choose the name of the env var using `with_env_var`.

In the meantime, I also tried to implement that kind of filtering using `dynamic_filter_fn` because there I have access to the `target`. But it is a mystery to me how I can test in that filter fn whether my `span_my_crate_trace` is a "parent" of the event in question.

**Reply by mladedav** [reply]
The [example](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/fn.dynamic_filter_fn.html#examples) seem to do exactly what you want where it checks for the current span's name. Or I maybe don't understand what you mean?

**Reply by dominikwerder** [reply]
Yes, I missed that I can access `current_span.parent()`.

Thank you!
